code/modules/admin/smites/puzzle.dm

```diff
@@ -0,0 +1,330 @@
+	var/first_tile_id = 1
+	var/other_tile_id = 2
+	if(empty_tile_id == 1 || empty_tile_id == 2) //Can't swap with empty one so just grab some in second row
+		first_tile_id = 4
+		other_tile_id = 5
+
+	var/turf/T1 = get_turf_for_id(first_tile_id)
+	var/turf/T2 = get_turf_for_id(other_tile_id)
+
+	var/obj/structure/puzzle_element/E1 = locate() in T1
+	var/obj/structure/puzzle_element/E2 = locate() in T2
+
+	E1.forceMove(T2)
+	E2.forceMove(T1)
+
+/proc/cmp_xy_desc(atom/movable/A,atom/movable/B)
+	if(A.y > B.y)
+		return -1
+	if(A.y < B.y)
+		return 1
+	if(A.x > B.x)
+		return 1
+	if(A.x < B.x)
+		return -1
+	return 0
+
+/obj/effect/sliding_puzzle/proc/elements_in_order()
+	return sortTim(elements,cmp=/proc/cmp_xy_desc)
+
+/obj/effect/sliding_puzzle/proc/get_base_icon()
+	var/icon/I = new('icons/obj/puzzle.dmi')
+	var/list/puzzles = icon_states(I)
+	var/puzzle_state = pick(puzzles)
+	var/icon/P = new('icons/obj/puzzle.dmi',puzzle_state)
+	return P
+
+/obj/effect/sliding_puzzle/proc/setup()
+	//First we slice the 96x96 icon into 32x32 pieces
+	var/list/puzzle_pieces = list() //id -> icon list
+
+	var/width = 3
+	var/height = 3
+	var/list/left_ids = list()
+	var/tile_count = width * height
+
+	//Generate per tile icons
+	var/icon/base_icon = get_base_icon()
+
+	for(var/id in 1 to tile_count)
+		var/y = width - round((id - 1) / width)
+		var/x = ((id - 1) % width) + 1
+
+		var/x_start = 1 + (x - 1) * world.icon_size
+		var/x_end = x_start + world.icon_size - 1
+		var/y_start = 1 + ((y - 1) * world.icon_size)
+		var/y_end = y_start + world.icon_size - 1
+
+		var/icon/T = new(base_icon)
+		T.Crop(x_start,y_start,x_end,y_end)
+		puzzle_pieces["[id]"] = T
+		left_ids += id
+
+	//Setup random empty tile
+	empty_tile_id = pick_n_take(left_ids)
+	var/turf/empty_tile_turf = get_turf_for_id(empty_tile_id)
+	empty_tile_turf.PlaceOnTop(floor_type,null)
+	var/mutable_appearance/MA = new(puzzle_pieces["[empty_tile_id]"])
+	MA.layer = empty_tile_turf.layer + 0.1
+	empty_tile_turf.add_overlay(MA)
+
+	elements = list()
+	var/list/empty_spots = left_ids.Copy()
+	for(var/spot_id in empty_spots)
+		var/turf/T = get_turf_for_id(spot_id)
+		T = T.PlaceOnTop(floor_type,null)
+		var/obj/structure/puzzle_element/E = new element_type(T)
+		elements += E
+		var/chosen_id = pick_n_take(left_ids)
+		E.puzzle_icon = puzzle_pieces["[chosen_id]"]
+		E.source = src
+		E.id = chosen_id
+		E.set_puzzle_icon()
+
+	if(!is_solvable())
+		make_solvable()
+
+/obj/structure/puzzle_element
+	name = "mysterious pillar"
+	desc = "puzzling..."
+	icon = 'icons/obj/artefacts.dmi'
+	icon_state = "puzzle_pillar"
+	anchored = FALSE
+	density = TRUE
+	var/id = 0
+	var/obj/effect/sliding_puzzle/source
+	var/icon/puzzle_icon
+
+/obj/structure/puzzle_element/Move(nloc, dir)
+	if(!isturf(nloc) ||  moving_diagonally || get_dist(get_step(src,dir),get_turf(source)) > 1)
+		return 0
+	else
+		return ..()
+
+/obj/structure/puzzle_element/proc/set_puzzle_icon()
+	cut_overlays()
+	if(puzzle_icon)
```

Comment:
early return here.

if(!puzzle_icon)
      return 
//Need to scale it down a bit to fit the static border
var/icon/C = new(puzzle_icon)